spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+    1. Run the following steps in |agent|:
+        1. If {{WindowSharedStorage/worklet}}'s [=module map=] is not [=map/empty=]:
+            1. Let |operationMap| be this {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
+            1. If |operationMap| [=map/contains=] |name|:
+                1.  Let |operation| be |operationMap|[|name|].
+                1. If |options| [=map/contains=] |data|:
+                    1. Let |argumentsList| be a new [=list=].
+                    1. [=list/Append=] |data| to |argumentsList|.
+                    1. [=Call=] |operation| with |argumentsList|.
+                1. Otherwise, [=call=] |operation| without any arguments list.
+        1. Resolve |promise| as undefined.
+    1. Return |promise|.
+  </div>
+
+  <div algorithm>
+    The <dfn method for="WindowSharedStorage">selectURL(|name|, |urls|, |options|)</dfn> method steps are:
+
+    1. Let |configPromise| be a new [=promise=].
+    1. If |urls| is empty or exceeds the maximum allowed length, reject |configPromise|.
+    1. Let |urlList| be an empty {{list}}.
+    1. For each |urlWithMetadata| in |urls|:
+        1. If |urlWithMetadata| has no field "`url`", reject |configPromise|.
+        1. Otherwise, let |urlString| be |urlWithMetadata|["`url`"].
+        1. Let |url| be the result of running a [=URL parser=] on |urlString|.
+        1. If |url| is not a valid {{URL}}, reject |configPromise|.
+        1. Otherwise, [=list/append=] |urlString| to |urlList|.
+        1. If |urlWithMetadata| has field "`reportingMetadata`":
+            1. Let |reportingMetadata| be |urlWithMetadata|["`reportingMetadata`"].
+            1. If |reportingMetadata| is invalid, reject |configPromise|.
+    1. If {{Worklet/addModule()}} has not yet been called, reject |configPromise|.
+    1. Let |fencedFrameConfig| be a {{FencedFrameConfig}}.
+    1. Resolve |configPromise| as |fencedFrameConfig|.
+    1. Let |indexPromise| be a new [=promise=].
+    1. Let |outsideSettings| be {{WindowSharedStorage/worklet}}'s [=relevant settings object=].
+    1. Let |agent| be the result of [=obtaining a worklet agent=] given |outsideSettings|.
+    1. Run the following steps in |agent|:
+        1. Let |index| be [=default index=].
+        1. If {{WindowSharedStorage/worklet}}'s [=module map=] is not [=map/empty=]:
+            1. Let |operationMap| be the associated {{SharedStorageWorkletGlobalScope}}'s [=SharedStorageWorkletGlobalScope/operation map=].
+            1. If |operationMap| [=map/contains=] |name|:
+                1.  Let |operation| be |operationMap|[|name|].
+                1. Let |argumentsList| be a new [=list=].
+                1. [=list/Append=] |urlList| to |argumentsList|.
+                1. If |options| [=map/contains=] |data|, [=list/Append=] |data| to |argumentsList|.
+                1. If |options| contains |data|, let |operationResult| be the result of running [=Call=] on |operation| with |argumentsList|.
```

Comment:
Is this the same if-statement as the line above?  Should we instead have one check and then two sub-steps?